engine/execution/computation/computer/computer.go

```diff
@@ -95,111 +104,156 @@ func (e *blockComputer) ExecuteBlock(
 }
 
 func (e *blockComputer) executeBlock(
-	ctx context.Context,
+	blockSpan opentracing.Span,
 	block *entity.ExecutableBlock,
 	stateView state.View,
 	programs *programs.Programs,
 ) (*execution.ComputationResult, error) {
 
 	blockCtx := fvm.NewContextFromParent(e.vmCtx, fvm.WithBlockHeader(block.Block.Header))
-
 	collections := block.Collections()
-
-	var gasUsed uint64
-
-	interactions := make([]*delta.SpockSnapshot, len(collections)+1)
-
-	events := make([]flow.Event, 0)
-	serviceEvents := make([]flow.Event, 0)
-	blockTxResults := make([]flow.TransactionResult, 0)
+	res := &execution.ComputationResult{
+		ExecutableBlock:    block,
+		Events:             make([]flow.Event, 0),
+		ServiceEvents:      make([]flow.Event, 0),
+		TransactionResults: make([]flow.TransactionResult, 0),
+	}
 
 	var txIndex uint32
+	var err error
+	var err1 error
+	var err2 error
+	var colView state.View
+	var prevColView state.View
+	var stateCommit flow.StateCommitment
 
-	for i, collection := range collections {
+	stateCommit = block.StartState
 
-		collectionView := stateView.NewChild()
+	if len(collections) > 0 {
 
-		e.log.Debug().
-			Hex("block_id", logging.Entity(block)).
-			Hex("collection_id", logging.Entity(collection.Guarantee)).
-			Msg("executing collection")
+		var wg sync.WaitGroup
 
-		collEvents, collServiceEvents, txResults, nextIndex, gas, err := e.executeCollection(
-			ctx, txIndex, blockCtx, collectionView, programs, collection,
-		)
-		if err != nil {
-			return nil, fmt.Errorf("failed to execute collection: %w", err)
+		// execute the first collection
+		colView = stateView.NewChild()
+		txIndex, err1 = e.executeCollection(blockSpan, txIndex, blockCtx, colView, programs, collections[0], res)
+		if err1 != nil {
+			return nil, fmt.Errorf("failed to execute collection: %w", err1)
+		}
+		prevColView = colView
+		err1 = stateView.MergeView(colView)
+		if err1 != nil {
+			return nil, fmt.Errorf("cannot merge view: %w", err1)
 		}
 
-		gasUsed += gas
-
-		txIndex = nextIndex
-		events = append(events, collEvents...)
-		serviceEvents = append(serviceEvents, collServiceEvents...)
-		blockTxResults = append(blockTxResults, txResults...)
-
-		interactions[i] = collectionView.(*delta.View).Interactions()
+		// executing collections
+		for _, collection := range collections[1:] {
+			wg.Add(2)
+			colView = stateView.NewChild()
+
+			go func() {
+				txIndex, err1 = e.executeCollection(blockSpan, txIndex, blockCtx, colView, programs, collection, res)
+				wg.Done()
+			}()
+
+			go func() {
+				stateCommit, err2 = e.commitView(blockSpan, prevColView, stateCommit, res)
+				wg.Done()
```

Comment:
I think that this would be more effective is we allow all `e.commitView` to be queued - we only need `stateCommit` to commit next view, not to execute collection. This will also ensure committing isn't slowed down by synchronization with execution

Reply:
not sure if i understand we need the views to commit which is the outcome of collection execution, i see your point if a collection execution is blocked by a states to he merged but in practice the commit delta time will be way smaller in time as soon as I make prove part run in parallel

Reply:
fixed